Lab2/AudioLab/AudioModel.m

```diff
@@ -0,0 +1,95 @@
+//
+//  AudioModel.m
+//  AudioLab
+//
+//  Created by Eric Smith on 9/12/18.
+//  Copyright © 2018 Eric Larson. All rights reserved.
+//
+
+#import "AudioModel.h"
+#import "Novocaine.h"
+#import "CircularBuffer.h"
+#import "FFTHelper.h"
+
+#define BUFFER_SIZE 2048*4
+
+@interface AudioModel ()
+@property (strong, nonatomic) AudioModel *singleton;
+@property (strong, nonatomic) Novocaine *audioManager;
+@property (strong, nonatomic) CircularBuffer *buffer;
+@property (strong, nonatomic) FFTHelper *fftHelper;
+@property (nonatomic) int bufferSize;
+@end
+
+@implementation AudioModel
+@synthesize singleton = _singleton;
+@synthesize bufferSize = _bufferSize;
+
++ (AudioModel *)sharedManager {
+    NSLog(@"Allocating shared model");
+    static AudioModel *sharedAudioModel = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedAudioModel = [[self alloc] initOnce];
+    });
+    
+    return sharedAudioModel;
+}
+
+- (AudioModel *)initOnce {
+    if (!_singleton) {
+        _singleton = [[AudioModel alloc] init];
+    }
+    return _singleton;
+}
+
+-(Novocaine*)audioManager{
+    if(!_audioManager){
+        _audioManager = [Novocaine audioManager];
+    }
+    return _audioManager;
+}
+
+
+-(int)getBufferSize {
+    return BUFFER_SIZE;
+}
+
+-(CircularBuffer*)buffer{
+    if(!_buffer){
+        _buffer = [[CircularBuffer alloc]initWithNumChannels:1 andBufferSize:BUFFER_SIZE];
+    }
+    return _buffer;
+}
+-(FFTHelper*)fftHelper{
+    if(!_fftHelper){
+        _fftHelper = [[FFTHelper alloc]initWithFFTSize:BUFFER_SIZE];
+    }
+    
+    return _fftHelper;
+}
+
+-(float*)getDataStream {
+    float* arrayData = malloc(sizeof(float)*BUFFER_SIZE);
+    [self.buffer fetchFreshData:arrayData withNumSamples:BUFFER_SIZE];
+    
+    return arrayData;
```

Comment:
I think arrayData needs to be a private member of the Model so that we can free this memory when the model is no longer needed. The malloc call can be moved to a lazy-getter and we can write a cleanup function that the view calls when it unloads.

Reply:
arrayData is now an instance variable of AudioModel. Clean up happens in the model's destructor. See: https://developer.apple.com/documentation/objectivec/nsobject/1571947-dealloc